p2p/peer.go

```diff
@@ -492,16 +571,47 @@ func (m *PeerManager) DialFailed(peerID NodeID, address PeerAddress) error {
 	defer m.mtx.Unlock()
 
 	delete(m.dialing, peerID)
+	for from, to := range m.upgrading {
+		if to == peerID {
+			// Unmark failed upgrade attempt.
+			delete(m.upgrading, from)
+		}
+	}
 
 	peer, err := m.store.Get(peerID)
 	if err != nil || peer == nil { // Peer may have been removed while dialing, ignore.
 		return err
 	}
-	if addressInfo := peer.LookupAddressInfo(address); addressInfo != nil {
-		addressInfo.LastDialFailure = time.Now().UTC()
-		addressInfo.DialFailures++
-		return m.store.Set(peer)
+	addressInfo := peer.LookupAddressInfo(address)
+	if addressInfo == nil {
+		return nil // Assume the address has been removed, ignore.
 	}
+	addressInfo.LastDialFailure = time.Now().UTC()
+	addressInfo.DialFailures++
+	if err = m.store.Set(peer); err != nil {
+		return err
+	}
+
+	// We spawn a goroutine that notifies DialNext() again when the retry
+	// timeout has elapsed, so that we can consider dialing it again.
+	//
+	// FIXME: We need to calculate the retry delay outside of the goroutine,
+	// since the arguments are currently pointers to structs shared in the
+	// peerStore. The peerStore should probably return struct copies instead,
+	// to avoid these sorts of issues.
+	if retryDelay := m.retryDelay(peer, addressInfo.DialFailures); retryDelay != time.Duration(math.MaxInt64) {
+		go func() {
+			timer := time.NewTimer(retryDelay)
+			defer timer.Stop()
+			select {
+			case <-timer.C:
```

Comment:
```suggestion
			case <-time.After(retryDelay):
```

Reply:
`time.After` will leak goroutines when `PeerManager.Close()` is called. I'll add a comment about this.